armory/data/utils.py

```diff
@@ -72,3 +76,54 @@ def verify_sha256(filepath: str, hash_value: str, block_size: int = 4096):
     value = sha256(filepath, block_size=block_size)
     if value != hash_value:
         raise ValueError(f"sha256 hash of {filepath}: {value} != {hash_value}")
+
+
+def verify_size(filepath: str, file_size: int):
+    size = os.path.getsize(filepath)
+    if size != file_size:
+        raise ValueError(f"file size of {filepath}: {size} != {file_size}")
+
+
+def download_verify_dataset_cache(dataset_dir, checksum_file):
+    with open(checksum_file, "r") as fh:
+        url, file_length, hash = fh.readline().strip().split()
+    # download
+    cache_dir = os.path.join(dataset_dir, "cache")
+    os.makedirs(cache_dir, exist_ok=True)
+    tar_filepath = os.path.join(cache_dir, os.path.basename(url))
+    if not os.path.exists(tar_filepath):
+        curl(url, dataset_dir, tar_filepath)
+
+    # verification
+    try:
+        verify_size(tar_filepath, int(file_length))
+        logger.info("Verifying sha256 hash of download...")
+        verify_sha256(tar_filepath, hash)
+    except ValueError:
+        os.remove(tar_filepath)
+        logger.info("Cached file download failed. Falling back to processing data...")
+        return
+
+    tmp_dir = os.path.join(
+        cache_dir,
+        "tmp_" + "".join(random.choice(string.ascii_lowercase) for _ in range(16)),
+    )
+    os.makedirs(tmp_dir)
+
+    logger.info("Extracting .tfrecord files from download...")
+    try:
+        with tarfile.open(tar_filepath, "r:gz") as tar_ref:
+            tar_ref.extractall(tmp_dir)
+    except tarfile.ReadError:
+        logger.info(f"Could not read tarfile: {tar_filepath}")
+        logger.info("Falling back to processing data...")
+        return
+    except tarfile.ExtractError:
+        logger.info(f"Could not extract tarfile: {tar_filepath}")
+        logger.info("Falling back to processing data...")
+        return
+
+    # move tmp_dir subdirectory to dataset_dir (tarball does not contain files at root level)
+    for directory in os.listdir(path=tmp_dir):
+        shutil.move(os.path.join(tmp_dir, directory), dataset_dir)
```

Comment:
You can run into an issue here if the top level directory exists (e.g., for a different version of the TFDS dataset), but not this specific one:
```
2020-03-19 23:42:27 5632ef0206c7 armory.data.utils[637] INFO Extracting .tfrecord files from download...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspace/armory/data/datasets.py", line 505, in librispeech_dev_clean
    subpath=os.path.join("plain_text", "1.1.0"),
  File "/workspace/armory/data/datasets.py", line 601, in _cache_dataset
    checksum_file=os.path.join(CACHED_CHECKSUMS_DIR, name + ".txt"),
  File "/workspace/armory/data/utils.py", line 128, in download_verify_dataset_cache
    shutil.move(os.path.join(tmp_dir, directory), dataset_dir)
  File "/opt/conda/lib/python3.7/shutil.py", line 564, in move
    raise Error("Destination path '%s' already exists" % real_dst)
shutil.Error: Destination path '/armory/datasets/librispeech_dev_clean_split' already exists
```

Since you know that the full directory path doesn't exist, I would suggest using os.makedirs to make all of the folders beneath that point and then moving over the leaf directory.

Reply:
Actually,  it is probably better to just find the first directory not in common and move that over, then cleanup the starting directory.

Reply:
If you call the following method instead of shutil.move, I think that will do the trick.
```python
import os                                                                                
import shutil                                                                            
                                                                                         
def move_merge(source='/armory/datasets/cache/tmp_esfupefxcusymnia/ucf101',              
        dest='/armory/datasets'):                                                        
    dest_location = os.path.join(dest, os.path.basename(source))                         
    if not os.path.exists(dest_location):                                                
        shutil.move(source, dest)                                                        
        return                                                                           
                                                                                         
    # Move up a directory, as shutil.move will error                                     
    filepaths = [os.path.join(source, x) for x in os.listdir(source) if not x.startswith(".")]
    if len(filepaths) != 1 or not os.path.isdir(filepaths[0]):                           
        raise ValueError(f"{source} not a single branch directory. Cannot recurse.")     
    move_merge(filepaths[0], dest_location)                                              
    os.rmdir(source)
```

Reply:
[source and dest are filled in there for clarity of my comment]